src/main/java/org/broadinstitute/hellbender/utils/nio/NioFileCopier.java

```diff
@@ -0,0 +1,624 @@
+
+            // Get our progress percentage:
+            final double rawProgressValuePercent = ((double) totalBytesRead / (double) srcFileSize) * 100.0;
+
+            // Round our progress to nearest PROGRESS_DISPLAY_PERCENT_INCREMENT:
+            final double progressValue = PROGRESS_DISPLAY_PERCENT_INCREMENT * (Math.floor(Math.abs(rawProgressValuePercent / PROGRESS_DISPLAY_PERCENT_INCREMENT)));
+
+            // Output our progress if we're ready for it:
+            if ( progressValue != lastProgressValue ) {
+
+                // Update our time:
+                final long   currentProgressTime_ns = System.nanoTime();
+                final long   dt_ms                  = (currentProgressTime_ns - lastProgressTime_ns) / NANOS_TO_MILLIS;
+                final double bytesPerMs             = ((double) progressBytesRead) / ((double) dt_ms);
+                lastProgressTime_ns = currentProgressTime_ns;
+
+                // Add the bytes/ms to our queue:
+                downloadBytesPerMilliSecond.add(bytesPerMs);
+
+                // Log our progress so far:
+                logProgress(progressValue, totalBytesRead, downloadBytesPerMilliSecond.stream().mapToDouble(x -> x).average().orElse(dt_ms));
+
+                // Get ready for the next call:
+                lastProgressValue = progressValue;
+                progressBytesRead = 0;
+            }
+        }
+    }
+
+    private void copyLoopWorker() {
+
+        try ( final InputStream inputStream = Files.newInputStream(getSource());
+              final OutputStream outputStream = Files.newOutputStream(getDest()) ){
+
+            // Get the file size of our source file:
+            srcFileSize = Files.size(getSource());
+            srcFileSizeNumDigits = (int)Math.ceil(Math.log10(srcFileSize));
+
+            if ( !isSilentCopy() ) {
+                logger.info("Initiating copy from " + getSource().toUri().toString() + " to " + getDest().toUri().toString());
+                logger.info("File size: " + srcFileSize + " bytes (" + FileUtils.byteCountToDisplaySize(srcFileSize) + ").");
+                logger.info("Please wait.  This could take a while...");
+            }
+
+            // Perform the copy:
+            while (true) {
+
+                // Read from our input:
+                final int bytesRead = inputStream.read(copyBuffer);
+                if ( bytesRead == -1 ) {
+                    break;
+                }
+
+                // Write to our output:
+                outputStream.write(copyBuffer, 0, bytesRead);
+
+                // Update our internal progress meter:
+                updateCopyProgress(bytesRead);
+            }
+        }
+        catch (final IOException ex) {
+            throw new GATKException("Could not copy file: " + source.toUri().toString() + " -> " + dest.toUri().toString(), ex);
+        }
+    }
+
+    /**
+     * Initiate the copy from {@link #source} to {@link #dest}.
+     */
+    public void initiateCopy() {
```

Comment:
I don't understand why this is separate from construction.  In all the cases you have, you immediately call this method after constructing the object.  It would make more sense to me to just do this as part of the construction.

Reply:
I don't like constructors with potentially huge side effects.  It seems better to me to be explicit on when the copy will happen.  It's not intuitively obvious that creating an object called `NioFileCopierWithProgressMeter` will kick off a copy.